File: 673/number-of-longest-increasing-subsequence.c
int findNumberOfLIS(int* nums, int numsSize){
    int table[numsSize]; //LIS的長度
    int combination[numsSize+1]; //LIS的組合數，包含i格的LIS有幾種可能
    int LISlen = 1;
    table[0] = 1;
    combination[0] = 1;
    for(int i=1; i<numsSize; i++){
        int temp=0;
        for(int k=0; k<i; k++){
            if(nums[k]<nums[i] && table[k]>temp) temp = table[k];
        }
        table[i] = temp +1;
        combination[i] = 0;
        for(int k=0; k<i; k++){
            if(table[k] == temp && nums[k]<nums[i]) combination[i] += combination[k];
        }
        if(combination[i]==0) combination[i] = 1;//若不是從前面推出來，自己有獨立1種解法

        if(table[i]>LISlen){
            LISlen = table[i];
        }
    }
    int ans=0;
    for(int i=0; i<numsSize; i++){
        if(table[i]==LISlen) ans+=combination[i];
    }
    return ans;
}//case 29/223: [1,2,4,3,5,4,7,2]

File: 673/number-of-longest-increasing-subsequence.py
class Solution:
    def findNumberOfLIS(self, nums: List[int]) -> int:
        N = len(nums)
        table = [0]*N # table[i] 包含 nums[i] 的 LIS 的長度
        count = [0]*N # count[i] 包含 nums[i] 的 LIS 的數量

        maxLIS = 0
        for i in range(N):
            table[i] = 1
            count[i] = 1
            for k in range(i):
                if nums[k] < nums[i] and table[k]+1 > table[i]:
                # 符合 LIS 的條件，而且更長，就整組更新
                    table[i] = table[k] + 1
                    count[i] = count[k]
                elif nums[k] < nums[i] and table[k]+1 == table[i]:
                # 符合 LIS 的條件，且一樣長，就增加count[i]
                    count[i] += count[k]
            if table[i] > maxLIS: # 順手比對，找出 maxLIS 的值
                maxLIS = table[i]

        ans = 0 # 含有 maxLIS 的 count[i] 總合
        for i in range(N):
            if table[i] == maxLIS: # nums[i] 符合 maxLIS 的長度
                ans += count[i] # 把數量累積
        # print(table)
        # print(count)
        return ans
